cityofaustin/atd-moped
# Adds council districts to project features and geography views

## Associated issues
- https://github.com/cityofaustin/atd-data-tech/issues/13311

This PR creates the mechanisms to associate projects with council districts. I'm looking forward to your feedback on this approach! Here are some implementation considerations:

1. I opted to persist council districts at the feature level instead of querying for them on the fly.

2. Because this approach persists district relationships at the feature level, we still need to do some additional computation/querying to surface council districts at the project and component level since one project has many components and one component has many features. This might warrant additional optimization.

Please also note that all of the changes to the moped editor are only temporary to help with testing. Viewing council district in the app is out of scope for this PR.

## Testing
**URL to test:** Local

**Steps to test:**

### First step: test the backfill

The migration is going to backfill council council districts for existing components, so you can test that by spinning up Moped at the beginning of this branch's history.

```shell
$ git fetch
$ git checkout 13311-jc-council-district-triggers
$ git checkout e28554e76aa6f5ddabf236af57677217833ddbe3
$ ./moped-database/hasura-cluster start
```

1. Navigate to the map tab of `Demo Project #1` and everything should look normal with two components on the map.

2. You need to manually fix the `features` table ID sequence because of a bug in the seed data. This fix is applied in my PR [here](https://github.com/cityofaustin/atd-moped/blob/b9b87bdb58351adb161b1620b12c6b847fba41a4/moped-database/seeds/1602292389297_initial_seed_staging.sql#L356-L360), but you have to do it manually because you already applied the old/buggy seeds when you started the cluster 👆 

```sql
# run this using your SQL client
SELECT
    setval('features_id_seq', (
            SELECT
                MAX(id)
                FROM features));
```

3. Now you will apply this branch's migrations.

```shell
# from atd-moped/moped-database
$ git switch - 
$ hasura migrate apply
$ hasura metadata apply
```

4. Refresh the map tab of `Demo Project #1` and observe that both components are listed as being in District 3. Nice! 

5. You can also use your SQL client or Hasura console to inspect the `project_geography` view, which has a `council_districts` column. 

### Second step: test cases

Now you will test a bunch of editing tasks while making sure that council district IDs are correct. While you do this, note that the district labels on the component map are placed on the district centroid, _so the label for district 9 is on the edge of district 1, and the label for district 1 is kinda hard to see_.

1. Create a new bike lane component. Select street segments within one council district. Save and confirm the district is listed correctly in the component list. Edit the bike lane and select additional segments in other districts.  Save and confirm the district is listed correctly in the component list.

2. Repeat those steps, but use line drawing instead/in addition to street segments.

3. Create a new Bike Box component. Select points in one or more districts, then save and confirm the districts are listed correctly. Repeat those steps using drawn points.

4. Lastly, try drawing a line and point-based component outside of any council district. Make sure that no district is associated with the component.

5. As one additional last step, test the updated seed data by completely restarting your cluster:

```shell
$ ./hasura-cluster restart
````


---
#### Ship list
